auth/src/source.rs

```diff
@@ -0,0 +1,474 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(dead_code)]
+
+use super::metadata;
+use crate::oauth2::{JwsClaims, JwsHeader};
+use crate::{AccessToken, Error, Result};
+use async_trait::async_trait;
+use chrono::{Duration, Utc};
+use rustls::sign::Signer;
+use rustls::sign::SigningKey;
+use rustls_pemfile::Item;
+use serde::{Deserialize, Serialize};
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+const DEFAULT_HEADER: JwsHeader = JwsHeader {
+    alg: "RS256",
+    typ: "JWT",
+    kid: None,
+};
+const DEFAULT_OAUTH_GRANT: &str = "urn:ietf:params:oauth:grant-type:jwt-bearer";
+const DEFAULT_USER_GRANT: &str = "refresh_token";
+
+/// An producer of a [AccessToken].
+#[async_trait]
+pub trait Source {
+    async fn token(&self) -> Result<AccessToken>;
+}
+
+/// Options for building a [ServiceAccountKeySource].
+pub struct ServiceAccountKeySourceBuilder<'a> {
+    file_path: Option<PathBuf>,
+    contents: Option<&'a [u8]>,
+
+    scopes: Vec<String>,
+}
+
+impl ServiceAccountKeySourceBuilder<'_> {
```

Comment:
I propose we write all new code without builders; which means dropping the builders from this file too.

I have clearly documented the negative effects of writing builders in an earlier conversation.  If there are benefits I haven't considered, please share them.